### PR TITLE
Marshalling interface generalized

### LINKED ISSUE #16: Generalize marshalling interface

### PR DESCRIPTION
**Pull request by nkaskov**
Closes #16 

Added three marshalling algorithms

- pack - From byteblob into some `X` type. `X` can be marshalling type or not. For example, in zk-marshalling `X` could be Redshift `proof_type`. 
- unpack - From `X` type into byteblob.
- repack - From `X` type into `Y` type conversion, 

All the algorithms are using only std::vector containers for blob input/output. It obviously has to be extended and generalized later.

As a subtask, very useful is_compatible traits were added. It let us to inference marshalling type needed for some arbitrary type. But it works only for array_list, integral and float_value for now.

### CODE REVIEW

File: include/nil/marshalling/algorithms/repack.hpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            status_type result_status;
+
+            std::vector<std::uint8_t> buffer = unpack<TInputEndian, 
+                std::uint8_t>(val, result_status);
+
+            TOutput result = pack<TOutputEndian, TOutput>(buffer, status);
+
+            status = status|result_status;
+            return result;
+        }
```
Comment (nemothenoone) [inline include/nil/marshalling/algorithms/repack.hpp:75]: Seems like it does use the type-to-type conversion mechanism.

File: include/nil/marshalling/algorithms/unpack.hpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            // static_assert(std::is_arithmetic<typename TContainer::value_type>::value);
+
+            using marshalling_type = typename is_compatible<TContainer>::template type<TEndian>;
+            using marshalling_internal_type = typename marshalling_type::element_type;
+
+            typename std::conditional<is_compatible<TContainer>::fixed_size, 
+                nil::marshalling::container::static_vector<marshalling_internal_type, marshalling_type::max_length()>, 
+                std::vector<marshalling_internal_type>>::type values;
+            for (const auto &val_i : val) {
+                values.push_back(marshalling_internal_type(val_i));
```
Comment (nemothenoone) [inline include/nil/marshalling/algorithms/unpack.hpp:141]: Can be replaced with `values.emplace_back(val_i);`

File: test/types.cpp [hunk 3]
```diff
@@ -50,8 +50,11 @@
 #include <nil/marshalling/compile_control.hpp>
 #include <nil/marshalling/units.hpp>
 #include <nil/marshalling/version.hpp>
+#include <nil/marshalling/inference.hpp>
 
 #include <nil/marshalling/algorithms/pack.hpp>
+#include <nil/marshalling/algorithms/disperse.hpp>
+#include <nil/marshalling/algorithms/marshal.hpp>
```
Comment (nemothenoone) [inline test/types.cpp:57]: Still kinda think this name (`marshal`) is questionable.

File: include/nil/marshalling/algorithms/repack.hpp (hunk 1 above)
Comment (nkaskov) [reply include/nil/marshalling/algorithms/repack.hpp:75]: Yep, it does. 

- We can substitute it later with old pack for some cases. But I decided not to do it right now. Old pack is to overcomplicated. If want to bring his logic in production and not just cut to some minimum, we need to do it more carefully.

- We anyway dont have such mechanism for more complicated cases, so it just implements some deafult behaviour.

**Review by nemothenoone — APPROVED** [review]

File: test/types.cpp (hunk 3 above)
Comment (nkaskov) [reply test/types.cpp]: Fixed.

File: include/nil/marshalling/algorithms/unpack.hpp (hunk 2 above)
Comment (nkaskov) [reply include/nil/marshalling/algorithms/unpack.hpp]: Thanks! Replaced.

**Pull request merged**